### PR TITLE
Fix calls to overrides of virtual mapped members

### LINKED ISSUE #3475: Query to Nullable Int(decimal, double...) does not work when raised via Dynamic.Linq.Core
I've the following Code:

	    [TestFixture]
	    public class Issue9999Tests : TestBase
	    {
		    internal class Linq2DbDynamicLinqCustomTypeProvider : DefaultDynamicLinqCustomTypeProvider
		    {
			    public override HashSet<Type> GetCustomTypes()
			    {
				    var types = base.GetCustomTypes();
				    types.Add(typeof(Sql));
				    return types;
			    }
		    }
    
		    [Table]
		    public class NumberLikeTestTable
		    {
			    [Column] public int?      IntNProp { get; set; }
		    }
    
		    public class NumberLikeTestObj
		    {
			    public NumberLikeTestTable? Obj { get; set; }
		    }
    
		    [Test]
		    public void NumberLikeTests([IncludeDataSources(TestProvName.AllSQLite)] string context)
		    {
			    ParsingConfig.Default.CustomTypeProvider = new Linq2DbDynamicLinqCustomTypeProvider();
    
			    using (var db = GetDataContext(context))
			    using (db.CreateLocalTable<NumberLikeTestTable>())
			    {
    #pragma warning disable CS8602
				    var query1 = db.GetTable<NumberLikeTestTable>().Select(x => new NumberLikeTestObj() { Obj = x }).Where(x => Sql.Like(x.Obj.IntNProp.ToString(), "1%")).Take(50);
				    var query2 = db.GetTable<NumberLikeTestTable>().Select(x => new NumberLikeTestObj() { Obj = x }).Where("((Sql.Like(it.Obj.IntNProp.ToString(),@0)))", new object[]{ "1%" }).Take(50);
    
				    var res1 = query1.ToList();
				    var res2 = query2.ToList();
    #pragma warning restore CS8602
			    }
		    }
	    }

```
Exception message:
LinqToDB.Linq.LinqException: "'Param_0.Obj.IntNProp.ToString()' cannot be converted to SQL."
```

The Debug View of the Expression:

Workin One (normal Linq):

		.Call System.Linq.Queryable.Take(
			.Call System.Linq.Queryable.Where(
				.Call System.Linq.Queryable.Select(
					.Call LinqToDB.DataExtensions.GetTable(.Constant<Tests.Model.TestDataConnection>(Tests.Model.TestDataConnection)),
					'(.Lambda #Lambda1<System.Func`2[Tests.UserTests.Issue9999Tests+NumberLikeTestTable,Tests.UserTests.Issue9999Tests+NumberLikeTestObj]>))
				,
				'(.Lambda #Lambda2<System.Func`2[Tests.UserTests.Issue9999Tests+NumberLikeTestObj,System.Boolean]>)),
			50)

		.Lambda #Lambda1<System.Func`2[Tests.UserTests.Issue9999Tests+NumberLikeTestTable,Tests.UserTests.Issue9999Tests+NumberLikeTestObj]>(Tests.UserTests.Issue9999Tests+NumberLikeTestTable $x)
		{
			.New Tests.UserTests.Issue9999Tests+NumberLikeTestObj(){
				Obj = $x
			}
		}

		.Lambda #Lambda2<System.Func`2[Tests.UserTests.Issue9999Tests+NumberLikeTestObj,System.Boolean]>(Tests.UserTests.Issue9999Tests+NumberLikeTestObj $x)
		{
			.Call LinqToDB.Sql.Like(
				.Call (($x.Obj).IntNProp).ToString(),
				"1%")
		}	

Broken one (created via dynamic linq):

		.Call System.Linq.Queryable.Take(
			.Call System.Linq.Queryable.Where(
				.Call System.Linq.Queryable.Select(
					.Call LinqToDB.DataExtensions.GetTable(.Constant<Tests.Model.TestDataConnection>(Tests.Model.TestDataConnection)),
					'(.Lambda #Lambda1<System.Func`2[Tests.UserTests.Issue9999Tests+NumberLikeTestTable,Tests.UserTests.Issue9999Tests+NumberLikeTestObj]>))
				,
				'(.Lambda #Lambda2<System.Func`2[Tests.UserTests.Issue9999Tests+NumberLikeTestObj,System.Boolean]>)),
			50)

		.Lambda #Lambda1<System.Func`2[Tests.UserTests.Issue9999Tests+NumberLikeTestTable,Tests.UserTests.Issue9999Tests+NumberLikeTestObj]>(Tests.UserTests.Issue9999Tests+NumberLikeTestTable $x)
		{
			.New Tests.UserTests.Issue9999Tests+NumberLikeTestObj(){
				Obj = $x
			}
		}

		.Lambda #Lambda2<System.Func`2[Tests.UserTests.Issue9999Tests+NumberLikeTestObj,System.Boolean]>(Tests.UserTests.Issue9999Tests+NumberLikeTestObj $var1)
		{
			.Call LinqToDB.Sql.Like(
				.Call (($var1.Obj).IntNProp).ToString(),
				"1%")
		}

### PR DESCRIPTION
**Pull request by jogibear9988**
fix #3475

### CODE REVIEW

File: Source/LinqToDB/Linq/Expressions.cs [hunk 1]
```diff
@@ -252,9 +252,13 @@ public static void SetGenericInfoProvider(Type type)
 		#region Public Members
 
 		static bool _checkUserNamespace = true;
+		static MethodInfo _toObjectToStringMethod = typeof(object).GetMethod("ToString")!;
```
Comment (MaceWindu) [inline Source/LinqToDB/Linq/Expressions.cs:255]: `methodinfo` should be stored in `Methods.cs` file

File: Tests/Linq/Tests.csproj [hunk 2]
```diff
@@ -15,6 +15,8 @@
 		<ProjectReference Include="..\FSharp\Tests.FSharp.fsproj" />
 
 		<PackageReference Include="NodaTime" />
+
+		<PackageReference Include="System.Linq.Dynamic.Core" />
```
Comment (MaceWindu) [inline Tests/Linq/Tests.csproj:19]: test dependencies should be in `linq2db.Tests.props`

File: Tests/Linq/UserTests/Issue3475Tests.cs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+		[Test]
+		public void NumberLikeTests([IncludeDataSources(TestProvName.AllSQLite)] string context)
+		{
+			ParsingConfig.Default.CustomTypeProvider = new Linq2DbDynamicLinqCustomTypeProvider();
+
+			using (var db = GetDataContext(context))
+			using (db.CreateLocalTable<NumberLikeTestTable>())
+			{
+#pragma warning disable CS8602
```
Comment (MaceWindu) [inline Tests/Linq/UserTests/Issue3475Tests.cs:46]: use nrt suppressions `!` instead of #pragma

File: Tests/Linq/UserTests/Issue3475Tests.cs [hunk 4]
```diff
@@ -0,0 +1,57 @@
+﻿#if NET5_0_OR_GREATER || NET472
```
Comment (MaceWindu) [inline Tests/Linq/UserTests/Issue3475Tests.cs:1]: why we need it?

Comment (jogibear9988) [reply Tests/Linq/UserTests/Issue3475Tests.cs:1]: not any more needed

File: Tests/Linq/Tests.csproj (hunk 2 above)
Comment (MaceWindu) [reply Tests/Linq/Tests.csproj]: because nobody reviewed it in time (fixed in v4 branch already)

**Comment by sdanyliv** [comment]
Still issue with `Expression.MapMember` will fix that later.

**Review by igor-tkachev — APPROVED** [review]

**Review by MaceWindu — APPROVED** [review]

**Comment by MaceWindu** [comment]
Don't merget yet, there are some issues with baselines. Looking into it

**Comment by MaceWindu** [comment]
Actually baselines are ok. `DateTimeFunctionsTests` show some places which could benefit from better date/time function mappings, but in general it's fine.

There are two informix test regressions I cannot reproduce locally, still looking into it

**Comment by MaceWindu** [comment]
solved, stale baselines

**Pull request merged**